pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -1186,59 +1203,37 @@ func (self *awsDisk) deleteVolume() (bool, error) {
 	return true, nil
 }
 
-// Gets the awsInstance for the EC2 instance on which we are running
-// may return nil in case of error
-func (s *AWSCloud) getSelfAWSInstance() (*awsInstance, error) {
-	// Note that we cache some state in awsInstance (mountpoints), so we must preserve the instance
-
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-
-	i := s.selfAWSInstance
-	if i == nil {
-		instanceId, err := s.metadata.GetMetadata("instance-id")
-		if err != nil {
-			return nil, fmt.Errorf("error fetching instance-id from ec2 metadata service: %v", err)
-		}
-		// privateDnsName, err := s.metadata.GetMetadata("local-hostname")
-		// See #11543 - need to use ec2 API to get the privateDnsName in case of private dns zone e.g. mydomain.io
-		instance, err := s.getInstanceByID(instanceId)
-		if err != nil {
-			return nil, fmt.Errorf("error finding instance %s: %v", instanceId, err)
-		}
-		privateDnsName := aws.StringValue(instance.PrivateDnsName)
-		availabilityZone, err := getAvailabilityZone(s.metadata)
-		if err != nil {
-			return nil, fmt.Errorf("error fetching availability zone from ec2 metadata service: %v", err)
-		}
-		instanceType, err := getInstanceType(s.metadata)
-		if err != nil {
-			return nil, fmt.Errorf("error fetching instance type from ec2 metadata service: %v", err)
-		}
-
-		i = newAWSInstance(s.ec2, instanceId, privateDnsName, availabilityZone, instanceType)
-		s.selfAWSInstance = i
+// Builds the awsInstance for the EC2 instance on which we are running.
+// This is called when the AWSCloud is initialized, and should not be called otherwise (because the awsInstance for the local instance is a singleton with drive mapping state)
+func (c *AWSCloud) buildSelfAWSInstance() (*awsInstance, error) {
+	if c.selfAWSInstance != nil {
+		panic("do not call buildSelfAWSInstance directly")
 	}
-
-	return i, nil
+	instanceId, err := c.metadata.GetMetadata("instance-id")
+	if err != nil {
+		return nil, fmt.Errorf("error fetching instance-id from ec2 metadata service: %v", err)
+	}
+	// privateDnsName, err := s.metadata.GetMetadata("local-hostname")
```

Comment:
Left this one because it's what we want to do, but doesn't work because of the issue...  As it's one line, I think I'll add a comment to say "we want to do this... <deadcode>" and then we'll flow into the "but we can't because of this issue".